### PR TITLE
OT275-25 : User register endpoint

### PR DESCRIPTION
**Pull request by r1ckt**
# [Ticket OT275-25](https://alkemy-labs.atlassian.net/browse/OT275-25)

### SUMMARY

- Added feature register user 

### HOW HAS THIS BEEN TESTED?

- [x] Postman.
- [x] Unit test.
- [ ] Integration test.
- [ ] No testing required (only applies for tech task).

**Screenshots**:

registerUserSuccess

BDUserEvidence

userAlreadyExists

exceptionsevidence

### CHECKLIST

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] I have added the new resource in the Postman Collection file.

### CODE REVIEW

**Comment by r1ckt** [comment]
fue sin querer lo del @ en los commits, no sabia que iba a pasar eso

File: src/main/java/com/alkemy/ong/infrastructure/database/repository/UserRepository.java [hunk 1]
```diff
@@ -22,4 +30,21 @@ public boolean exists(Identifiable<Long> identifiable) {
...
+  @Override
+  public Optional<User> find(String email) {
+    return userSpringRepository.findByEmail(email).map(userEntityMapper::toDomain);
+  }
+
+  @Override
+  public User add(User user) {
+    UserEntity userEntity = userEntityMapper.toEntity(user);
+    String roleName = user.getRole().getName();
+    userEntity.setRole(getRoleEntity(roleName));
```
Comment (mkain0) [inline src/main/java/com/alkemy/ong/infrastructure/database/repository/UserRepository.java:42]: Esto no es responsabilidad del add de UserRepository. Es responsabilidad del service.

File: src/test/java/com/alkemy/ong/application/service/CreateUserUseCaseServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  @Test
+  void shouldSaveUserWhenUserDoesNotExist() {
+    given(userRepository.find(EMAIL)).willReturn(Optional.empty());
+
+    Role role = new Role();
+    role.setName(ROLE);
+
+    User user = new User();
+    user.setEmail(EMAIL);
+    user.setRole(role);
```
Comment (mkain0) [inline src/test/java/com/alkemy/ong/application/service/CreateUserUseCaseServiceTest.java:54]: Cuando el user ingresa al service no tiene role seteado.

**Review by mkain0 — CHANGES_REQUESTED** [review]

**Review by mkain0 — APPROVED** [review]

File: src/main/java/com/alkemy/ong/application/service/CreateUserUseCaseService.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+public class CreateUserUseCaseService implements ICreateUserUseCase {
+  private final IUserRepository userRepository;
+
+  @Override
+  public User add(User user) {
+    if (userRepository.find(user.getEmail()).isPresent()) {
+      throw new UserAlreadyExists(ErrorMessage.USER_ALREADY_EXISTS.getMessage());
+    }
+    com.alkemy.ong.domain.Role role = new com.alkemy.ong.domain.Role();
+    role.setName(Role.USER.toString());
```
[the remark this answers is not in the document]
Comment (r1ckt) [reply src/main/java/com/alkemy/ong/application/service/CreateUserUseCaseService.java]: Lo hice asi porque en la entidad tenia @Column(length = 5) y no me persistia el nombre completo del rol, deberia realizar cambios en la entity?

**Pull request merged**